NitroxClient/GameLogic/Spawning/Metadata/Extractor/GenericEntityMetadataExtractor.cs

```diff
@@ -0,0 +1,63 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using NitroxModel.DataStructures.GameLogic.Entities.Metadata;
+using NitroxModel.DataStructures.Util;
+using UnityEngine;
+
+namespace NitroxClient.GameLogic.Spawning.Metadata.Extractor;
+
+public abstract class GenericEntityMetadataExtractor<I, O> : EntityMetadataExtractor where O : EntityMetadata
+{
+    public abstract O Extract(I entity);
+
+    public override Optional<EntityMetadata> From(object o)
+    {
+        EntityMetadata result = Extract((I)o);
+
+        return Optional.Of(result);
+    }
+}
+
+public abstract class EntityMetadataExtractor
+{
+    public abstract Optional<EntityMetadata> From(object o);
+
+    private static readonly Dictionary<Type, EntityMetadataExtractor> processors;
+
+    static EntityMetadataExtractor()
+    {
+        processors = Assembly.GetExecutingAssembly()
+                             .GetTypes()
+                             .Where(t => typeof(EntityMetadataExtractor).IsAssignableFrom(t) &&
+                                    t.IsClass && !t.IsAbstract)
+                             .Select(Activator.CreateInstance)
+                             .Cast<EntityMetadataExtractor>()
+                             .ToDictionary(p => p.GetType().BaseType.GetGenericArguments()[0]);    
```

Comment:
Is there a reason you are using plain reflection and not Autofac?

Reply:
swapped to autofac